rust/x509/src/x509_example_rust/src/wasm_only.rs

```diff
@@ -0,0 +1,509 @@
+            let response = cdk_ecdsa::ecdsa_public_key(args)
+                .await
+                .map_err(|e| format!("ecdsa_public_key failed {}", e.1))?;
+            response.0.public_key
+        }
+    };
+
+    // try to initialize the cache with the fetched public key or returne the
+    // cached value, because we were making an async call between the cache
+    // check and cache initialization
+    Ok(ROOT_CA_PUBLIC_KEY.with(move |inner| inner.get_or_init(|| result).clone()))
+}
+
+async fn pem_certificate_signed_by_root_ca<Signer>(
+    profile: Profile,
+    serial_number: SerialNumber,
+    validity: Validity,
+    subject: Name,
+    subject_public_key_info: SubjectPublicKeyInfoOwned,
+    signer: Signer,
+) -> Result<String, String>
+where
+    Signer: Sign,
+    Signer: Keypair + DynSignatureAlgorithmIdentifier,
+    Signer::VerifyingKey: EncodePublicKey,
+{
+    let mut builder = CertificateBuilder::new(
+        profile,
+        serial_number,
+        validity,
+        subject,
+        subject_public_key_info,
+        &signer,
+    )
+    .expect("Create certificate");
+
+    let blob = builder
+        .finalize()
+        .map_err(|e| format!("failed to finalize certificate builder: {e:?}"))?;
+
+    let signature = BitString::from_bytes(&signer.sign(&blob).await?)
+        .map_err(|e| format!("wrong signature length: {e:?}"))?;
+
+    let certificate = builder
+        .assemble(signature)
+        .map_err(|e| format!("failed to assemble certificate: {e:?}"))?;
+
+    certificate
+        .to_pem(LineEnding::LF)
+        .map_err(|e| format!("failed to encode certificate: {e:?}"))
+}
+
+fn verify_certificate_request_signature(certificate_request: &CertReq) -> Result<(), String> {
+    fn certificate_as_message_and_public_key_bytes(
+        certificate_request: &CertReq,
+    ) -> (Vec<u8>, Vec<u8>) {
+        let mut message = vec![];
+        certificate_request
+            .info
+            .encode(&mut message)
+            .expect("failed to encode certificate request info");
+
+        let public_key_bytes = certificate_request
+            .info
+            .public_key
+            .subject_public_key
+            .raw_bytes()
+            .to_vec();
+
+        (message, public_key_bytes)
+    }
+    let result = match &certificate_request.algorithm {
+        AlgorithmIdentifier::<der::Any> {
+            oid: ed25519::pkcs8::ALGORITHM_OID,
+            parameters: None,
+        } => {
+            let (message, public_key_bytes) =
+                certificate_as_message_and_public_key_bytes(certificate_request);
+            verify_ed25519_signature(
+                certificate_request.signature.raw_bytes(),
+                message.as_slice(),
+                public_key_bytes.as_slice(),
+            )
+        }
+        AlgorithmIdentifier::<der::Any> {
+            oid: ecdsa::ECDSA_SHA256_OID,
+            parameters: None, // secp256k1 is non-standard and is sometimes encoded as None
```

Comment:
What software are you seeing that omits the parameters? That seems pretty surprising / non-conforming.

Reply:
`openssl` in `rust`

Reply:
A generated CSR looks like this:
```
Certificate Request:
    Data:
        Version: 1 (0x0)
        Subject: CN=Test corporation, O=Test Inc, C=US
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:9c:1e:84:0e:50:ec:8b:dc:c0:45:6f:0e:94:7e:
                    7b:57:ea:bb:e8:14:76:25:aa:bf:6f:eb:f1:86:a7:
                    d6:ae:14:49:b7:1d:b4:c8:15:96:93:e5:18:ce:d4:
                    d1:65:46:41:b6:92:05:07:13:b8:6f:bd:64:ba:21:
                    91:b1:21:59:a4
                ASN1 OID: secp256k1
        Attributes:
            (none)
            Requested Extensions:
    Signature Algorithm: ecdsa-with-SHA256
    Signature Value:
        30:45:02:21:00:90:bc:77:24:55:da:a7:20:3b:e4:98:2d:ab:
        a1:4e:0f:b5:91:ff:87:27:00:36:7e:ce:b6:3b:38:16:ac:2f:
        cf:02:20:70:c1:67:b7:46:15:b8:3c:f4:57:42:42:87:1a:b4:
        ee:09:01:a0:26:70:86:7c:ee:59:ad:80:a2:b7:51:e0:9e
-----BEGIN CERTIFICATE REQUEST-----
MIHzMIGaAgEAMDsxGTAXBgNVBAMMEFRlc3QgY29ycG9yYXRpb24xETAPBgNVBAoM
CFRlc3QgSW5jMQswCQYDVQQGEwJVUzBWMBAGByqGSM49AgEGBSuBBAAKA0IABJwe
hA5Q7IvcwEVvDpR+e1fqu+gUdiWqv2/r8Yan1q4USbcdtMgVlpPlGM7U0WVGQbaS
BQcTuG+9ZLohkbEhWaSgADAKBggqhkjOPQQDAgNIADBFAiEAkLx3JFXapyA75Jgt
q6FOD7WR/4cnADZ+zrY7OBasL88CIHDBZ7dGFbg89FdCQocatO4JAaAmcIZ87lmt
gKK3UeCe
-----END CERTIFICATE REQUEST-----
```

Reply:
I was thinking about the wrong field here, it is correct/expected that the parameters are omitted here.

RFC 5758

>  When the ecdsa-with-SHA224, ecdsa-with-SHA256, ecdsa-with-SHA384, or ecdsa-with-SHA512 algorithm identifier appears in the algorithm field as an AlgorithmIdentifier, the encoding MUST omit the parameters field.

Reply:
Thanks for pointing that out! I removed the parameters from the code.